server/src/main/java/com/linecorp/centraldogma/server/internal/httpapi/ContentServiceV1.java

```diff
@@ -0,0 +1,457 @@
+
+            final long commitTimeMillis = System.currentTimeMillis();
+            final CommitMessageDto commitMessage = commitMessageAndChange.getKey();
+            final CompletableFuture<Revision> resultRevisionFuture = changesFuture.thenCompose(
+                    changes -> push(commitTimeMillis, author, repository, normalizedRevision,
+                                    commitMessage, changes.values()));
+
+            final EntryType type = entryTypeFromChange(change);
+            return resultRevisionFuture.thenApply(result -> DtoConverter.convert(
+                    result, projectName, repoName, change.path(), type, commitTimeMillis));
+        });
+    }
+
+    private static Entry<CommitMessageDto, Change<?>> commitMessageAndChange(AggregatedHttpMessage message) {
+        final JsonNode jsonNode = getJsonNode(message);
+        checkArgument(jsonNode.get("path") != null && jsonNode.get("commitMessage") != null &&
+                      jsonNode.get("content") != null, "need a path, a content and a commit message to commit");
+
+        final JsonNode contentNode = jsonNode.get("content");
+        final String path = jsonNode.get("path").textValue();
+        final Change<?> change = getChange(contentNode, path);
+        final CommitMessageDto commitMessage = convertCommitMessage(jsonNode.get("commitMessage"));
+        return Maps.immutableEntry(commitMessage, change);
+    }
+
+    private static Change<?> getChange(JsonNode contentNode, String path) {
+        if (contentNode.getNodeType() == JsonNodeType.STRING) {
+            validateFilePath(path, "path");
+            return Change.ofTextUpsert(path, contentNode.textValue());
+        }
+
+        validateJsonFilePath(path, "path");
+        return Change.ofJsonUpsert(path, contentNode);
+    }
+
+    private static CommitMessageDto convertCommitMessage(JsonNode jsonNode) {
+        final CommitMessageDto commitMessage = Jackson.convertValue(jsonNode, CommitMessageDto.class);
+        checkArgument(!isNullOrEmpty(commitMessage.summary()), "summary should be non-null");
+        return commitMessage;
+    }
+
+    private static EntryType entryTypeFromChange(Change<?> change) {
+        return change.type().contentType().isAssignableFrom(JsonNode.class) ? EntryType.JSON : EntryType.TEXT;
+    }
+
+    private CompletionStage<Revision> push(long commitTimeMills, Author author, Repository repository,
+                                           Revision revision, CommitMessageDto commitMessage,
+                                           Iterable<Change<?>> changes) {
+        final String summary = commitMessage.summary();
+        final String detail = commitMessage.detail();
+        final Markup markup = commitMessage.markup();
+
+        final CompletableFuture<Revision> normalizeFuture = repository.normalize(revision);
+        final CompletableFuture<Revision> headFuture = repository.normalize(Revision.HEAD);
```

Comment:
If `revision` is `HEAD` then we should not call `normalize` twice, because two `normalize(HEAD)` calls may return different revisions on contention.